### PR TITLE
feat: add option skip malformed parts

### PR DESCRIPTION
**Pull request by dmytrokasianenko-outreach**
This PR is proposal for the feature which allows to skip parts that can't be parsed. It won't return error, just add them to "problems"

This can be useful if you need to parse everything what is possible to parse and turned off by default.

For example, we have emails which have body of the email starting at the beginning of the part (without headers). Currently parser returns error,  but it can be beneficial to skip this part.

```
...
Content-Type: multipart/report; report-type=delivery-status;
	boundary="12345/example.com"
Auto-Submitted: auto-generated (failure)
MIME-Version: 1.0

--12345/example.com
[EXTERNAL EMAIL]

The original message was received
...
```

### CODE REVIEW

**Comment by jhillyerd** [comment]
I'm in favor of this, will take a bit of time to review, and may hold off merging until after the next release.

**Comment by iredmail** [comment]
How about add a new ReadEnvelopeWithOptions() instead of modifying existing ReadEnvelope()?

I'd like to add options like:

- skip reading / parsing mail body
- Skip (base64) decoding attachments

**Comment by jhillyerd** [comment]
My initial plan for options was like @iredmail describes above, however I think if I designed it today, I'd have a options builder struct with ReadEnvelope/etc as methods on it.  This would make it easier for people to mock in tests.  The original package funcs could just forward to those methods on a default options object.

That being said, I haven't thought about the strategy @dmytrokasianenko-outreach used before, so I will spend some time on it.  Probably not this weekend though.  Most code would continue to work with that change, but I think I've seen breakage from that type of change in the past.  I expect we'd want to go from 0.9 series to 0.10 or 1.0 in that case.

**Comment by dmytrokasianenko-outreach** [comment]
Probably better solutions could be to use option interfaces, see  [Uber Go Style Guide for functional options]( https://github.com/uber-go/guide/blob/master/style.md#functional-options). Mainly:

> Note that there's a method of implementing this pattern with closures but we believe that the pattern above provides more flexibility for authors and is easier to debug and test for users. In particular, it allows options to be compared against each other in tests and mocks, versus closures where this is impossible. Further, it lets options implement other interfaces, including fmt.Stringer which allows for user-readable string representations of the options.

WDYT?

**Comment by kadukf-outreach** [comment]
Q1: @[jhillyerd](https://github.com/jhillyerd) so that it would be possible to replace some part of the message parsing flow? So that we will have open/close principle in place - open for extension but closed for modifications - closed part would be message parsing orchestration (method ReadParts) which drives boundary reader but the particular steps would be replaceable (use defaults it nothing provided), right? IMO the question is how much of the openings is needed/desired.

Q2: would it be possible to introduce custom error for malformed parts so that when processing "problems" it would be possible to get a pointer to the malformed part, type of issue (header, content, etc).

Q3: Shouldn't explicitly "skip malformed parts" be part of the default behavior so that after parsing we get some results with references to the failed parts so that we can handle it accordingly, wdyt?

**Comment by jhillyerd** [comment]
> Probably better solutions could be to use option interfaces, see [Uber Go Style Guide for functional options](https://github.com/uber-go/guide/blob/master/style.md#functional-options). Mainly:
> 
> > Note that there's a method of implementing this pattern with closures but we believe that the pattern above provides more flexibility for authors and is easier to debug and test for users. In particular, it allows options to be compared against each other in tests and mocks, versus closures where this is impossible. Further, it lets options implement other interfaces, including fmt.Stringer which allows for user-readable string representations of the options.
> 
> WDYT?

After giving it some thought, I'm OK with the functional options pattern you are using, and it seems more acceptable in Go than a builder pattern.  What I'm not OK with is forcing our users to build their options from scratch for every call to enmime, most users will be using identical options for every parse.  Thoughts on making config public (while keeping the fields private)?

**Comment by dmytrokasianenko-outreach** [comment]
> What I'm not OK with is forcing our users to build their options from scratch for every call to enmime, most users will be using identical options for every parse. Thoughts on making config public (while keeping the fields private)?

I have a idea, but it would require API changes:
```go
// build parser once, with all configs
p := enmime.NewParser(ops...)
// use multiple times
p.Parse(...)
p.Parse(...)
```

**Comment by iredmail** [comment]
> > What I'm not OK with is forcing our users to build their options from scratch for every call to enmime, most users will be using identical options for every parse. Thoughts on making config public (while keeping the fields private)?
> 
> I have a idea, but it would require API changes:
> 
> ```go
> // build parser once, with all configs
> p := enmime.NewParser(ops...)
> // use multiple times
> p.Parse(...)
> p.Parse(...)
> ```

A new `ReadEnvelopeWithOptions()` is better IMO, no backward-comptability concern.

**Comment by jhillyerd** [comment]
> I have a idea, but it would require API changes

I like that.  We could modify the existing ReadEnvelope + ReadParts to use a default set of options, so the old API continues to work.

**Comment by dmytrokasianenko-outreach** [comment]
> A new ReadEnvelopeWithOptions() is better IMO, no backward-comptability concern.

function will still require to pass options and build config every time you want to parse an email.

**Comment by dmytrokasianenko-outreach** [comment]
@jhillyerd updated code to use functional options as interfaces and added parser constructor which consumes options

**Review by jhillyerd — CHANGES_REQUESTED** [review]
Overall this looks great, I added one question and a couple changes.

My plan is to merge this into a separate branch, release the final 0.9.x, and then merge this to master in the next week or so.  This will kick off 0.10, which I hope will be a rather short series of releases before going 1.0.  :rocket:

File: options.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package enmime
+
+// Option to configure parsing.
+type Option interface {
+	apply(p *Parser)
+}
+
+// SkipMalformedParts sets parsing to skip parts that's can't be parsed.
+func SkipMalformedParts() Option {
```
[the remark this answers is not in the document]
Comment (dmytrokasianenko-outreach) [reply options.go:9]: yep, you are correct. To me `func WithCache(c bool)` is a little bit confusing and it is not clear what is default behavior - would it cache by default or not.

[the remark this answers is not in the document]
Comment (jhillyerd) [reply options.go:9]: ...and while I think it's extremely unlikely that we'd change the default for SkipMalformedParts, every PR is going to copy this option, and I will need to have this conversation many times over. 😆

File: part_test.go [hunk 2]
```diff
@@ -374,6 +374,57 @@ func TestMultiMixedParts(t *testing.T) {
...
+		Parent:      test.PartExists,
+		ContentType: "text/plain",
+		Charset:     "us-ascii",
+		PartID:      "3",
+	}
+	test.ComparePart(t, p, wantp)
+
+	want = "Section two"
+}
+
```
[the remark this answers is not in the document]
Comment (dmytrokasianenko-outreach) [reply part_test.go:428]: test added

File: options.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jhillyerd) [reply options.go:9]: Yeah, let's add the bool to make this a template for others to copy.  It's looks less elegant than your way, but easier for me in the long wrong.

**Review by jhillyerd — APPROVED** [review]

**Comment by jhillyerd** [comment]
Thanks!  This is probably one of the most exciting changes for enmime in years.  I'll try to release 0.9.x & merge it to master this week.

**Comment by jhillyerd** [comment]
Merged to master

**Pull request merged**